### PR TITLE
Allow negative heights in RPC calls

### PR DESCRIPTION
**Pull request by oxarbitrage**
For issue https://github.com/zcash/zcash/issues/2197

Currently adding the feature to `getblock` and `getblockhash`. There is another candidate: `getblocksubsidy` however i want to have some review about these 2 first before repeating what could be a bad approach.

### CODE REVIEW

**Comment by daira** [comment]
Concept ACK. Please make clear in the doc precisely which block is referred to by -1 (I think it's the last known valid block on the chain).

**Comment by oxarbitrage** [comment]
@daira :  A height of -1 is going to dump results for last valid block - 1.

I updated the docs to reflect this at https://github.com/zcash/zcash/pull/4280/commits/574963ef8724d43e3ab1b8b59fb75a799ff18eec

**Comment by daira** [comment]
Hmm, that convention means there's no index that always references the last valid block.

**Comment by oxarbitrage** [comment]
Yes that is correct. Will you prefer that -1 indicate last valid block ? I can try to update the code to that.

**Comment by oxarbitrage** [comment]
Changed the convention as suggested @daira . There is a pretty big benefit being able to get the last valid block with `-1` for the clients,  it seems they cant do it now with just 1 call (need to do `getblockcount` then `getblock`). Thanks for the heads-up.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -587,6 +587,11 @@ UniValue getblockhash(const UniValue& params, bool fHelp)
     LOCK(cs_main);
 
     int nHeight = params[0].get_int();
+
+    if(nHeight < 0) {
```
Comment (daira) [inline src/rpc/blockchain.cpp]:
```suggestion
    if (nHeight < 0) {
```

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -706,7 +711,7 @@ UniValue getblock(const UniValue& params, bool fHelp)
     // If height is supplied, find the hash
     if (strHash.size() < (2 * sizeof(uint256))) {
         // std::stoi allows characters, whereas we want to be strict
-        regex r("[[:digit:]]+");
+        regex r("-?[[:digit:]]+");
```
Comment (daira) [inline src/rpc/blockchain.cpp]: This allows `-0` (interpreted as `0`, i.e. the genesis block). Is that intended?

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -719,9 +724,14 @@ UniValue getblock(const UniValue& params, bool fHelp)
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block height parameter");
         }
 
+        if(nHeight < 0) {
```
Comment (daira) [inline src/rpc/blockchain.cpp]:
```suggestion
        if (nHeight < 0) {
```

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (oxarbitrage) [reply src/rpc/blockchain.cpp]: Its unintended. Thanks.

I tried to find something better online but no luck so built this one:

`(?:(-?)[1-9][0-9]*|[0-9]+)`

Where:

- `?:` - if
- `(-?)` - Condition, string starting with `-` character.
- `[1-9]` - Then next character must be a digit between 1 and 9.
- `[0-9]*` - Can be more digits from 0 to 9.
- `|` - Else
- `[0-9]+` One or more 0 to 9 digits.

https://github.com/zcash/zcash/pull/4280/commits/3fb8af512c464140d6ca67c814313bd72bb17780

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (oxarbitrage) [reply src/rpc/blockchain.cpp]: Sorry about that ... again. I changed it in the context of https://github.com/zcash/zcash/commit/3fb8af512c464140d6ca67c814313bd72bb17780

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (oxarbitrage) [reply src/rpc/blockchain.cpp]: changed at https://github.com/zcash/zcash/commit/3fb8af512c464140d6ca67c814313bd72bb17780

**Comment by oxarbitrage** [comment]
@daira maybe can re review with the modified regex ? thanks.

**Comment by daira** [comment]
@zkbot r+

**Comment by daira** [comment]
This is a deterministic failure. I suspect there's a test that was checking negative block numbers are invalid, that needs updating.

**Comment by daira** [comment]
@zkbot retry

**Comment by daira** [comment]
@zkbot r+

**Pull request merged**